### DISCUSSION TITLE
Getting the basename of packageFile?

### DISCUSSION THREAD

**Discussion by vsimon** [renovatebot/renovate#13791] [Request Help]
Not sure if it's possible to even do this, I'd like to use the basename of packageFile in the additionalBranchPrefix or commitMessagePrefix fields.

I'm using environments with helmfile and I have a values folder containing different helm chart versions for each environment.

```
helmfiles.yaml
values
- dev.yaml
- prod.yaml
```

Example values/dev.yaml:
```
# renovate: depName=kubernetes-dashboard registryUrl=https://kubernetes.github.io/dashboard
version: 5.0.5
```

And a renovate.json containing:

```
  "helmfile": {
    "enabled": false
  },
  "regexManagers": [
    {
      "fileMatch":[
        "\\.yaml$"
      ],
      "matchStrings":[
        "depName=(?<depName>.*?) registryUrl=(?<registryUrl>.*?)\\sversion: (?<currentValue>.*)"
      ],
      "datasourceTemplate": "helm"
    }
  ],
  "additionalBranchPrefix": "{{ packageFile }}-",
  "commitMessagePrefix": "{{ packageFile }}:"
```

The branch it wants to create is `branch=renovate/releases/kubernetes-dashboard/values/dev.yaml-kubernetes-dashboard-5.x`. I see there is a parentDir field but I'd have to put each file in a dir of the same name like `values/dev/dev.yaml`. Is there any way to get like the basename(packageFile) so that the branch would be `renovate/releases/kubernetes-dashboard/dev-kubernetes-dashboard-5.x`?

**Comment by viceice** [comment]
`packageFile`? 

https://docs.renovatebot.com/templates/

**Comment by rarkins** [comment]
Are you just wanting to remove the file extension (in this case `.yaml`)?

You could submit a PR for this (e.g. name: `packageFileBaseName`, or you can probably achieve the same using the [`replace`](https://docs.renovatebot.com/templates/#replace) extension we enabled for handlebars.

**Comment by vsimon** [comment]
Thanks @viceice @rarkins , Yes, I was trying out `{{ packageFile }}` first.

Yes, I am wanting to remove the file extension but also the folders leading up to the file. Like if my file structure was arbitrarily `foo/bar/baz/dev.yaml`, I'd like to have `dev: Update chart...` in the commit message instead of `foo/bar/baz/dev.yaml: Update chart...`.

Thanks for the suggest, I'll try the `replace` next! i.e. `{{{ replace 'values/', '.yaml', packageFile }}}`

**Comment by vsimon** [comment]
Hi @viceice @rarkins, following the replace documentation: https://docs.renovatebot.com/templates/#replace

This fails (with the commas between arguments like in the documentation):

`"additionalBranchPrefix": "{{{replace 'values/', '', packageFile}}}-"`

With error:

```
"loggerErrors": [
  {
    "name": "renovate",
    "level": 50,
    "logContext": "maoxqZ4WRMeGIfz3rRiy6",
    "repository": "...",
    "err": {
      "message": "Parse error on line 1:\n.../{{{replace 'values', '', packageFile}}}\n-----------------------^\nExpecting 'CLOSE_RAW_BLOCK', 'CLOSE', 'CLOSE_UNESCAPED', 'OPEN_SEXPR', 'CLOSE_SEXPR', 'ID', 'OPEN_BLOCK_PARAMS', 'STRING', 'NUMBER', 'BOOLEAN', 'UNDEFINED', 'NULL', 'DATA', got 'INVALID'",
      "stack": "Error: Parse error on line 1:\n.../{{{replace 'values', '', packageFile}}}\n-----------------------^\nExpecting 'CLOSE_RAW_BLOCK', 'CLOSE', 'CLOSE_UNESCAPED', 'OPEN_SEXPR', 'CLOSE_SEXPR', 'ID', 'OPEN_BLOCK_PARAMS', 'STRING', 'NUMBER', 'BOOLEAN', 'UNDEFINED', 'NULL', 'DATA', got 'INVALID'\n    at Parser.parseError (/usr/src/app/node_modules/handlebars/lib/handlebars/compiler/parser.js:199:11)\n    at Parser.parse (/usr/src/app/node_modules/handlebars/lib/handlebars/compiler/parser.js:251:22)\n    at parseWithoutProcessing (/usr/src/app/node_modules/handlebars/lib/handlebars/compiler/base.js:24:20)\n    at HandlebarsEnvironment.parse (/usr/src/app/node_modules/handlebars/lib/handlebars/compiler/base.js:30:13)\n    at compileInput (/usr/src/app/node_modules/handlebars/lib/handlebars/compiler/compiler.js:532:19)\n    at ret (/usr/src/app/node_modules/handlebars/lib/handlebars/compiler/compiler.js:546:18)\n    at Object.compile (/usr/src/app/node_modules/renovate/lib/util/template/index.ts:196:38)\n    at generateBranchName (/usr/src/app/node_modules/renovate/lib/workers/repository/updates/branch-name.ts:90:34)\n    at applyUpdateConfig (/usr/src/app/node_modules/renovate/lib/workers/repository/updates/flatten.ts:59:21)\n    at flattenUpdates (/usr/src/app/node_modules/renovate/lib/workers/repository/updates/flatten.ts:126:28)\n    at branchifyUpgrades (/usr/src/app/node_modules/renovate/lib/workers/repository/updates/branchify.ts:21:19)\n    at lookup (/usr/src/app/node_modules/renovate/lib/workers/repository/process/extract-update.ts:113:36)\n    at extractDependencies (/usr/src/app/node_modules/renovate/lib/workers/repository/process/index.ts:105:11)\n    at Object.renovateRepository (/usr/src/app/node_modules/renovate/lib/workers/repository/index.ts:39:52)\n    at Object.start (/usr/src/app/node_modules/renovate/lib/workers/global/index.ts:126:7)\n    at /usr/src/app/node_modules/renovate/lib/renovate.ts:16:22"
    },
    "msg": "Repository has unknown error"
  }
```

This works (sans commas):

`"additionalBranchPrefix": "{{{replace 'values/' '' packageFile}}}-"`

Possibly a documentation mismatch.

---

Going one step further, this also fails (I kinda didn't expect it to work because of the nested nature):

`"additionalBranchPrefix": "{{{replace '\\.yaml' '' {{{replace 'values/' '' packageFile}}}}}}-"`

Because I wanted to replace both the `.yaml` extension and the folders leading up to the packageFile. I think I will file a PR then, as a workaround doesn't seem possible with more than one replace.

**Reply by viceice** [reply]
@HonkingGoose can you please fix the docs, comma's need to be removed.

The replace supports Regex, so you can maybe use match groups in replacement text.

https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/String/replace#switching_words_in_a_string

**Reply by vsimon** [reply]
Thanks!

**Reply by vsimon** [reply]
Thanks, I'll try the match groups suggest !

**Reply by HonkingGoose** [reply]
> @HonkingGoose can you please fix the docs, comma's need to be removed.

See pull request #13813. 😉

**Reply by vsimon** [reply]
Thanks so much @viceice for the fantastic help! I believe it works!

The solution ended up at:

```
"additionalBranchPrefix": "{{{replace 'values/(.*)\\.yaml' '$1' packageFile}}}-"
```

**Reply by viceice** [reply]
Maybe we should add this as an example to the docs too.

**Comment by vsimon** [comment]
Based off this subcomment: [#13791 (reply in thread)](https://github.com/renovatebot/renovate/discussions/13791#discussioncomment-2045123)

The solution ended up at:

```
"additionalBranchPrefix": "{{{replace 'values/(.*)\\.yaml' '$1' packageFile}}}-"
```
